### PR TITLE
Fix problems with wrapped errors in logrus

### PR DESCRIPTION
**Pull request by aaronlehmann**
This vendors a fork of logrus with fixes to deal with errors generated by the `github.com/pkg/errors` package. Without these fixes, the errors are mishandled when passed to `WithError`:
- When printing to a terminal (colors enabled), stack traces are included in the error fields on a log line. Example output:

```
DEBU[0000] failed to load local CA certificate           error=local root CA certificate does not exist
github.com/docker/swarmkit/ca.init
    /Users/diogo/go/src/github.com/docker/swarmkit/ca/certificates.go:74
github.com/docker/swarmkit/agent.init
    /Users/diogo/go/src/github.com/docker/swarmkit/agent/worker.go:369
main.init
    /Users/diogo/go/src/github.com/docker/swarmkit/cmd/swarmd/main.go:199
runtime.main
    /usr/local/opt/go/libexec/src/runtime/proc.go:172
runtime.goexit
    /usr/local/opt/go/libexec/src/runtime/asm_amd64.s:2086 module=node/tls
```
- When printing to a file (or with colors disabled), the error value is
  missing entirely.

The upstream PR is https://github.com/Sirupsen/logrus/pull/419

I think it's important that we fix this somehow before 1.13, since we've started using `github.com/pkg/errors` quite a bit in Docker and SwarmKit. This PR is intended as a placeholder until the PR is merged upstream, and to allow testing of the upstream PR. If the upstream PR isn't merged soon, we should continue by vendoring the fork, or find another solution.

cc @diogomonica @stevvooe @tonistiigi @cpuguy83 @thaJeztah

### CODE REVIEW

**Comment by diogomonica** [comment]
FORK!

**Comment by tonistiigi** [comment]
LGTM, I'd be ok with merging fork until this is patched in upstream. For example https://jenkins.dockerproject.org/job/Docker%20Master%20(arm)/3283/console is flaky but logs are useless because error is empty.

**Comment by thaJeztah** [comment]
I pinged him on the PR, perhaps my magic touch helps speeding it up :smile:

**Comment by stevvooe** [comment]
If I remember correctly, I think you can actually inject a textformatter with this fix, rather than fork the package.

**Comment by aaronlehmann** [comment]
Haven't heard anything from upstream, unfortunately.

I looked into @stevvooe's suggestion of using a custom formatter, and it looks like this would work. It doesn't feel ideal because we'd have to put a fixed version of the formatter somewhere in Docker's source code (where?). SwarmKit and other projects would also need a copy of this formatter to work around the bugs. In some ways it feels cleaner just to vendor the fork.

How do people want to move forward with this?

**Comment by aaronlehmann** [comment]
Rebased.

Unfortunately, there has been no response from logrus upstream in 3 weeks, so I think it's time to move forward with this fork. Hopefully it will be a temporary fork.

**Comment by thaJeztah** [comment]
Agreed; let's move forward with the fork as temporary solution until the PR is merged upstream

**Comment by thaJeztah** [comment]
ping @stevvooe PTAL

File: vendor/src/github.com/Sirupsen/logrus/writer.go [hunk 1]
```diff
@@ -7,18 +7,40 @@ import (
...
 func (logger *Logger) Writer() *io.PipeWriter {
+	return logger.WriterLevel(InfoLevel)
+}
+
+func (logger *Logger) WriterLevel(level Level) *io.PipeWriter {
 	reader, writer := io.Pipe()
 
-	go logger.writerScanner(reader)
+	var printFunc func(args ...interface{})
+	switch level {
```
Comment (stevvooe) [inline vendor/src/github.com/Sirupsen/logrus/writer.go]: How did these end up "re-leveled"? Seems odd to be doing dispatch at this point. How was this handled before?

Comment (aaronlehmann) [inline vendor/src/github.com/Sirupsen/logrus/writer.go]: This isn't related to any changes I made. It comes from rebasing off the latest upstream version of logrus.

I can base the changes on the version we had vendored previously if you preferred, but I thought working from the latest upstream would be best.

**Comment by stevvooe** [comment]
LGTM on the fork

Reviewing parts of the changes to fork here. Wondering if we should carry forward with something more context integrated (see swarmkit/log and distribution/context for examples of that direction) if we are getting upstream changes from `logrus`.

**Comment by aaronlehmann** [comment]
Rebased again

**Pull request merged**